enhancements/proposals/1978-scylladbdatacenter-api/README.md

```diff
@@ -0,0 +1,764 @@
+Even the simplest rules require providing 10'ish lines with 4'ish indentation levels.
+This makes the entire resource huge and UX isn't great.
+To improve it, the new field `topologyLabelSelector` added on the datacenter and rack level,
+provides a label selector which will be translated into placement nodeAffinity rules selecting nodes matching these labels.
+Label selectors from the datacenter and rack level will be merged into the final node affinity. 
+In case of a collision on keys, rack ones take precedence.
+
+To prevent misuse of this simplification keys of the provided label selector will be validated against the `topology.kubernetes.io/*` rule,
+and usage of both `nodeAffinity` and `topologyLabelSelector` will be forbidden by the webhook.
+Users wanting to construct more complex placement rules will have to do it using `placement` spec.
+
+Given following spec:
+```yaml
+spec:
+  topologyLabelSelector:
+    topology.kubernetes.io/region: us-east1
+  racks:
+  - name: a
+    topologyLabelSelector:
+      topology.kubernetes.io/zone: us-east1a
+```
+
+Results in rack `a` having `nodeAffinity`:
+```yaml
+nodeAffinity:
+  requiredDuringSchedulingIgnoredDuringExecution:
+    nodeSelectorTerms:
+      - matchExpressions:
+        - key: topology.kubernetes.io/region
+          operator: In
+          values:
+            - us-east1
+        - key: topology.kubernetes.io/zone
+          operator: In
+          values:
+          - us-east1a
+```
+
+### Uniform datacenter
+
+ScyllaDB recommendation is to have uniform racks. 
+I propose to extend the API so that the racks can be defined uniformly in a simplified manner, 
+while preserving the existing flexibility of having non-uniform racks by overriding it with dedicated specifications, if needed.
+
+#### Rack template
+
+New field `rackTemplate` on datacenter level, allows defining shared properties across all racks.
+Racks inherit these properties unless the same is specified on rack level, then overwritten value takes precedence.
+
+#### Node count
+
+A new field `nodes` will be added on the datacenter level. Scylla Operator will balance this number between all defined racks.
+Having a single place where the desired number of nodes is specified, enables us to have support for scale subresource,
+which brings users handy scaling operations using `kubectl scale` and paves a path towards available Kubernetes autoscalers.
+The downside of it is that these won't be able to add more racks, but provides a nice way of scaling for static topology clusters.
+
+Datacenter level `nodes` cannot be used together with `nodes` specified per rack. Webhook will prevent it.
+
+The balancing algorithm is iterative, adding/removing one node at a time in each iteration to ensure that racks stay as balanced as possible.
+
+```go
+func balanceNodesAcrossRacks(racks []*appsv1.StatefulSet, totalNodes int) []*appsv1.StatefulSet {
```

Comment:
Why do we need this?

Making this depend on the current state kind of beats the point of indempotent desired configs. The racks are ordered and they should always have their scale grown so that `nodes(N)>=nodes(N+1)` which will produce the same desired sub-scales, no matter the state.

Reply:
> they should always have their scale grown so that nodes(N)>=nodes(N+1)

I don't think that's a sufficient condition. If you want to distribute 7 nodes over 3 racks both [3,3,1] and [3,2,2] satisfy it, but the latter would be desired, I suppose. I.e. you also want to minimise the pair-wise difference of rack sizes imo.
Having said that, this algorithm seems way over-complicated. Isn't it just `racks[i] = N/len(racks) + (i < N mod len(racks))`?

Reply:
you are right that I've left the definition a bit ambiguous (I have talked about at most 1 difference somewhere else), it should have been `nodes(N+1) <= nodes(N) <= nodes(N+1)+1` - effectively you'd distribute the node count to each rack  by 1 node, in order, cyclically

Reply:
here's the code in go https://go.dev/play/p/w0lZMWN_w-N since that notation wasn't clear

Reply:
to merge the ideas above into an algorithm, how about this?
```golang
for i := range sd.spec.racks {
    sd.spec.racks[i].nodes = len(sd.spec.nodes) / len(sd.spec.racks)
    if i < len(sd.spec.nodes) % len(sd.spec.racks) {
        sd.spec.racks[i].nodes++
    }
}
```

Reply:
> Why do we need this?
> Making this depend on the current state kind of beats the point of indempotent desired configs. The racks are ordered and they should always have their scale grown so that nodes(N)>=nodes(N+1) which will produce the same desired sub-scales, no matter the state.

Relying on existing state is important because when recommended NetworkTopologyStrategy is used by the keyspace, Scylla tries to put a copy of entire dataset on every rack. Which means, scaling down the rack when cluster is growing can cause that rack won't have enough capacity to hold the copy of the data. 

> here's the code in go https://go.dev/play/p/w0lZMWN_w-N since that notation wasn't clear

Your algorithm produce `[3,2,2]` for nodes=7, racks=3, while `[2,2,2,1]` for nodes=7, racks=4 which is not what we would want, as when cluster is growing, it shouldn't scale down any rack due to above.

What my algorithm would produce is `[3,2,2,0]` which is what I think is expected in this case. 

---

Another important topic we haven't though about is decommissioning a single rack. Right now we forbid removing a rack until it's first scaled down to 0. With balancing algorithm and inability to override desired node count for particular rack this seems impractical.

Given that we disagree about the desired behavior for scaling up, and we haven't though well entire topology change topic I propose to push this feature out of this proposal and prepare next one focusing on this particular change. 
We should consult Core team about what should we do in these cases, describe and understand every scenario.

Reply:
> Which means, scaling down the rack when cluster is growing can cause that rack won't have enough capacity to hold the copy of the data.

The algorithm only computes the desired config, it doesn't say how you apply it. We may first scale the racks that need it up, wait and then scale down, right?

> Your algorithm produce [3,2,2] for nodes=7, racks=3, while [2,2,2,1] for nodes=7, racks=4 which is not what we would want, as when cluster is growing, it shouldn't scale down any rack due to above.

> What my algorithm would produce when you already have [3,2,2] cluster is [3,2,2,0] which is what I think is expected in this case.

How would you get into the desired state if you want to say grow your RF from 3 to 5 by declaring 2 extra racks. If they become of size 0, it can't server RF=5.

> Another important topic we haven't though about is decommissioning a single rack. Right now we forbid removing a rack until it's first scaled down to 0. With balancing algorithm and inability to override desired node count for particular rack this seems impractical.

That's a good point. Since we know the existing racks and the desired ones, I guess we could scale it on behind, before pruning it out. A validation may be an easier start - if you use global "nodes", you can only add racks. (You can switch it to manual nodes and then remove it.)

> Given that we disagree about the desired behavior for scaling up

well, we now know where the difference is - I'll setup a call today and we can talk about the approach or deffer it, if we still have doubts

Reply:
> here's the code in go https://go.dev/play/p/w0lZMWN_w-N since that notation wasn't clear

I now see you were faster :)

Reply:
This about this a second time, this may well deserve a separate proposal with a discussion limited for that topic. We know the implication for the initial API - the `nodes` counts have to be pointers but required for now, so we can later loosen this.

Reply:
> Your algorithm produce [3,2,2] for nodes=7, racks=3, while [2,2,2,1] for nodes=7, racks=4 which is not what we would want, as when cluster is growing, it shouldn't scale down any rack due to above.
What my algorithm would produce when you already have [3,2,2] cluster is [3,2,2,0] which is what I think is expected in this case.

From the perspective of just this simple API, i.e. only declaring the number of nodes and number of racks, I think you'd expect that increasing the number of racks would result in redistributing the nodes. Whether and how that plays with Scylla is an entirely different topic, and maybe a good example for the `nodes` field being too much of a simplification?

> Given that we disagree about the desired behavior for scaling up, and we haven't though well entire topology change topic I propose to push this feature out of this proposal and prepare next one focusing on this particular change.
We should consult Core team about what should we do in these cases, describe and understand every scenario.

+1 for moving this to a separate proposal, there's more to discuss here. I'd also remove the `.spec.nodes` from this proposal entirely in case we end up with something more complicated.